File: framework/src/play/src/main/java/play/server/ApplicationProvider.java
/*
 * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
 */

package play.server;

import play.Application;
import play.mvc.Http;
import play.mvc.Result;
import scala.compat.java8.OptionConverters;

import java.util.Optional;

/**
 * Provides information about a Play Application running inside a Play server.
 */
public class ApplicationProvider {

    private final Application application;
    private final play.core.ApplicationProvider underlying;

    public ApplicationProvider(Application application) {
        this.application = application;
        this.underlying = play.core.ApplicationProvider$.MODULE$.apply(application.asScala());
    }

    /**
     * @return The Scala version of this application provider.
     */
    public play.core.ApplicationProvider asScala() {
        return underlying;
    }

    /**
     * @return Returns an Optional with the application, if available.
     */
    public Optional<Application> get() {
        return Optional.ofNullable(application);
    }

    /**
     * Handle a request directly, without using the application.
     * @param requestHeader the request made.
     */
    public Optional<Result> handleWebCommand(Http.RequestHeader requestHeader) {
        return OptionConverters
                .toJava(this.underlying.handleWebCommand(requestHeader.asScala()))
                .map(play.api.mvc.Result::asJava);
    }
}

File: core/play/src/main/scala/play/core/ApplicationProvider.scala
/*
 * Copyright (C) Lightbend Inc. <https://www.lightbend.com>
 */

package play.core

import java.io._
import scala.util.Try
import scala.util.Success

import play.api._
import play.api.mvc._

/**
 * Provides source code to be displayed on error pages
 */
trait SourceMapper {
  def sourceOf(className: String, line: Option[Int] = None): Option[(File, Option[Int])]

  def sourceFor(e: Throwable): Option[(File, Option[Int])] = {
    e.getStackTrace.find(element => sourceOf(element.getClassName).isDefined).flatMap { interestingStackTrace =>
      sourceOf(interestingStackTrace.getClassName, Option(interestingStackTrace.getLineNumber))
    }
  }
}

/**
 * Provides information about a Play Application running inside a Play server.
 */
trait ApplicationProvider {

  /**
   * Get the application. In dev mode this lazily loads the application.
   *
   * NOTE: This should be called once per request. Calling multiple times may result in multiple compilations.
   */
  def get: Try[Application]

  /**
   * Handle a request directly, without using the application.
   */
  @deprecated("This method is no longer called; WebCommands are now handled by the DefaultHttpRequestHandler", "2.7.0")
  def handleWebCommand(requestHeader: play.api.mvc.RequestHeader): Option[Result] = None
}

object ApplicationProvider {

  /**
   * Creates an ApplicationProvider that wraps an Application instance.
   */
  def apply(application: Application) = new ApplicationProvider {
    val get: Try[Application] = Success(application)
  }
}

trait HandleWebCommandSupport {
  def handleWebCommand(
      request: play.api.mvc.RequestHeader,
      buildLink: play.core.BuildLink,
      path: java.io.File
  ): Option[Result]
}
